File: tests/aat/common/matcher/tvlp.py
from expects import expect, be_a, be_empty, be_none, be_above
from expects.matchers import Matcher

from common.matcher import be_valid_dynamic_results

import client.models


class _be_valid_tvlp_configuration(Matcher):
    def _match(self, config):
        expect(config).to(be_a(client.models.TvlpConfiguration))
        expect(config.id).not_to(be_empty)
        expect(config.state).not_to(be_none)
        expect(config.time).not_to(be_none)
        expect(config.error).to(be_none)
        expect(config.profile).to(be_a(client.models.TvlpProfile))
        return True, ['is valid TVLP configuration']


class _be_valid_block_tvlp_profile(Matcher):
    def _match(self, profile):
        expect(profile).to(be_a(client.models.TvlpProfileBlock))
        expect(profile.series).not_to(be_empty)
        for conf in profile.series:
            expect(conf).to(be_a(client.models.TvlpProfileBlockSeries))
            expect(conf.length).not_to(be_none)
            expect(conf.length).to(be_above(0))
            expect(conf.resource_id).not_to(be_none)
            expect(conf.config).to(be_a(client.models.BlockGeneratorConfig))
        return True, ['is valid Block TVLP profile']


class _be_valid_memory_tvlp_profile(Matcher):
    def _match(self, profile):
        expect(profile).to(be_a(client.models.TvlpProfileMemory))
        expect(profile.series).not_to(be_empty)
        for conf in profile.series:
            expect(conf).to(be_a(client.models.TvlpProfileMemorySeries))
            expect(conf.length).not_to(be_none)
            expect(conf.length).to(be_above(0))
            expect(conf.config).to(be_a(client.models.MemoryGeneratorConfig))
        return True, ['is valid Memory TVLP profile']


class _be_valid_cpu_tvlp_profile(Matcher):
    def _match(self, profile):
        expect(profile).to(be_a(client.models.TvlpProfileCpu))
        expect(profile.series).not_to(be_empty)
        for conf in profile.series:
            expect(conf).to(be_a(client.models.TvlpProfileCpuSeries))
            expect(conf.length).not_to(be_none)
            expect(conf.length).to(be_above(0))
            expect(conf.config).to(be_a(client.models.CpuGeneratorConfig))
        return True, ['is valid Cpu TVLP profile']


class _be_valid_packet_tvlp_profile(Matcher):
    def _match(self, profile):
        expect(profile).to(be_a(client.models.TvlpProfilePacket))
        expect(profile.series).not_to(be_empty)
        for conf in profile.series:
            expect(conf).to(be_a(client.models.TvlpProfilePacketSeries))
            expect(conf.length).not_to(be_none)
            expect(conf.length).to(be_above(0))
            expect(conf.target_id).not_to(be_none)
            expect(conf.config).to(be_a(client.models.PacketGeneratorConfig))
        return True, ['is valid Packet TVLP profile']


class _be_valid_tvlp_result(Matcher):
    def _match(self, result):
        expect(result).to(be_a(client.models.TvlpResult))
        expect(result.id).not_to(be_none)
        expect(result.tvlp_id).not_to(be_none)
        for prof in [result.memory, result.block, result.cpu]:
            if prof == None: continue
            for gen_result in prof:
                expect(gen_result.dynamic_results).to(be_valid_dynamic_results)
        return True, ['is valid TVLP result']


be_valid_tvlp_configuration = _be_valid_tvlp_configuration()
be_valid_block_tvlp_profile = _be_valid_block_tvlp_profile()
be_valid_memory_tvlp_profile = _be_valid_memory_tvlp_profile()
be_valid_cpu_tvlp_profile = _be_valid_cpu_tvlp_profile()
be_valid_packet_tvlp_profile = _be_valid_packet_tvlp_profile()
be_valid_tvlp_result = _be_valid_tvlp_result()

File: src/swagger/converters/tvlp.cpp

#include "tvlp.hpp"

#include "packet_generator.hpp"
#include "swagger/v1/model/TvlpConfiguration.h"

namespace swagger::v1::model {

void from_json(const nlohmann::json& j, TvlpProfile_cpu_series& cpu_series)
{
    auto val = const_cast<nlohmann::json&>(j);

    cpu_series.setLength(val.at("length"));

    auto gc = std::make_shared<CpuGeneratorConfig>();
    gc->fromJson(val.at("config"));
    cpu_series.setConfig(gc);
}

void from_json(const nlohmann::json& j, TvlpProfile_cpu& cpu_profile)
{
    auto val = const_cast<nlohmann::json&>(j);

    cpu_profile.getSeries().clear();
    for (auto& item : val["series"]) {
        auto newItem = std::make_shared<TvlpProfile_cpu_series>(
            item.get<TvlpProfile_cpu_series>());
        cpu_profile.getSeries().push_back(newItem);
    }
}

void from_json(const nlohmann::json& j, TvlpProfile_block_series& block_series)
{
    auto val = const_cast<nlohmann::json&>(j);

    block_series.setResourceId(val.at("resource_id"));
    block_series.setLength(val.at("length"));

    auto gc = std::make_shared<BlockGeneratorConfig>();
    gc->fromJson(val.at("config"));
    block_series.setConfig(gc);
}

void from_json(const nlohmann::json& j, TvlpProfile_block& block_profile)
{
    auto val = const_cast<nlohmann::json&>(j);

    block_profile.getSeries().clear();
    for (auto& item : val["series"]) {
        auto newItem = std::make_shared<TvlpProfile_block_series>(
            item.get<TvlpProfile_block_series>());
        block_profile.getSeries().push_back(newItem);
    }
}

void from_json(const nlohmann::json& j,
               TvlpProfile_memory_series& memory_series)
{
    auto val = const_cast<nlohmann::json&>(j);

    memory_series.setLength(val.at("length"));

    auto gc = std::make_shared<MemoryGeneratorConfig>();
    gc->fromJson(val.at("config"));
    memory_series.setConfig(gc);
}

void from_json(const nlohmann::json& j, TvlpProfile_memory& memory_profile)
{
    auto val = const_cast<nlohmann::json&>(j);

    memory_profile.getSeries().clear();
    if (val.find("series") != val.end()) {
        for (auto& item : val["series"]) {
            auto newItem = std::make_shared<TvlpProfile_memory_series>(
                item.get<TvlpProfile_memory_series>());
            memory_profile.getSeries().push_back(newItem);
        }
    }
}

void from_json(const nlohmann::json& j,
               TvlpProfile_packet_series& packet_series)
{
    auto val = const_cast<nlohmann::json&>(j);

    packet_series.setTargetId(val.at("target_id"));
    packet_series.setLength(val.at("length"));

    auto gc = std::make_shared<PacketGeneratorConfig>();
    *gc = j["config"].get<PacketGeneratorConfig>();
    packet_series.setConfig(gc);
}

void from_json(const nlohmann::json& j, TvlpProfile_packet& packet_profile)
{
    auto val = const_cast<nlohmann::json&>(j);

    packet_profile.getSeries().clear();
    if (val.find("series") != val.end()) {
        for (auto& item : val["series"]) {
            auto newItem = std::make_shared<TvlpProfile_packet_series>(
                item.get<TvlpProfile_packet_series>());
            packet_profile.getSeries().push_back(newItem);
        }
    }
}

void from_json(const nlohmann::json& j, TvlpProfile& profile)
{
    auto val = const_cast<nlohmann::json&>(j);
    if (val.find("memory") != val.end()) {
        if (!val["memory"].is_null()) {
            auto newItem = std::make_shared<TvlpProfile_memory>(
                val.at("memory").get<TvlpProfile_memory>());
            profile.setMemory(newItem);
        }
    }

    if (val.find("block") != val.end()) {
        if (!val["block"].is_null()) {
            auto newItem = std::make_shared<TvlpProfile_block>(
                val.at("block").get<TvlpProfile_block>());
            profile.setBlock(newItem);
        }
    }

    if (val.find("cpu") != val.end()) {
        if (!val["cpu"].is_null()) {
            auto newItem = std::make_shared<TvlpProfile_cpu>(
                val.at("cpu").get<TvlpProfile_cpu>());
            profile.setCpu(newItem);
        }
    }

    if (val.find("packet") != val.end()) {
        if (!val["packet"].is_null()) {
            auto newItem = std::make_shared<TvlpProfile_packet>(
                val.at("packet").get<TvlpProfile_packet>());
            profile.setPacket(newItem);
        }
    }
}

void from_json(const nlohmann::json& j, TvlpConfiguration& generator)
{
    if (j.find("id") != j.end()) generator.setId(j.at("id"));

    if (j.find("time_scale") != j.end())
        generator.setTimeScale(j.at("time_scale"));
    else
        generator.setTimeScale(1.0);

    if (j.find("load_scale") != j.end())
        generator.setLoadScale(j.at("load_scale"));
    else
        generator.setLoadScale(1.0);

    generator.setProfile(
        std::make_shared<TvlpProfile>(j.at("profile").get<TvlpProfile>()));
}

} // namespace swagger::v1::model